Numark Mixtrack Pro 3 mapping

djsteph opened the pull request:
Hello 
I have completed the Numark Mixtrack Pro 3 mapping files (version 1.1). Changes noted in the JS file. Forum thread is here: http://mixxx.org/forums/viewtopic.php?f=7&t=7286&start=140#p28103 and documentation of mapping is here: http://mixxx.org/wiki/doku.php/numark_mixtrack_pro_3

Thank you 

Steph

djsteph committed:
Numark Mixtrack Pro 3 mapping

Numark Mixtrack Pro 3 mapping version 1.1

Be-ing commented:
Great to see this finally submitted :) It may take me a bit to get a chance to thoroughly review this because there are still a few other mappings I haven't finished reviewing. At first glace the code and wiki both look pretty good.

sblaisot commented:
PLease update windows uninstaller: https://mixxx.org/wiki/doku.php/contributing_mappings#windows_installer_update

djsteph committed:
Updated uninstaller as per comment for Mixtrack Pro 3 mapping

sblaisot commented on build/nsis/Mixxx.nsi:
```diff
@@ -546,6 +546,9 @@ Section "Uninstall"
   Delete "$INSTDIR\controllers\Xone K2.midi.xml"
   Delete "$INSTDIR\controllers\Xone-K2-scripts.js"
+  Delete "$INSTDIR\controllers\Numark-Mixtrack-3.midi.xml"
+  Delete "$INSTDIR\controllers\Numark-Mixtrack-3-scripts.js"
```
Please keep all files in alphabetical order. This will be easier for future maintenance.
Could you please move these two lines in the correct place ?

djsteph committed:
Update to ensure controller is listed in alphabetical order

Be-ing commented:
Okay, I read through the wiki. I did quite a bit of editing to make the language and formatting more concise and easier to navigate.

It's good that you've included some options, but the amount of options is kinda overwhelming. Some of them could be removed. IMO the iCutEnabled and fastSeekEnabled options are unnecessary. Because these features have to be deliberately activated by the user by holding Shift, they do not interfere with other functionality. If a user doesn't want to use them, all they have to do is not use them. Keeping the options available clutters the script and wiki.
Also the TapExpandLibrary option could be removed if pushing the browse encoder is mapped to [Master]. maximize_library. I think that would be a better use of it considering that there are load buttons right next to it.
I think it would be clearer to rename PADLoopButtonPressed to PADLoopButtonHold and PADSampleButtonPressed to PADSampleButtonHold.

What color are the pads in Manual Loop Mode?

Is the microphone input mixed with the main output in hardware? Is the input available as a sound card input to the computer?

Are there any volume controls available in your OS mixer program?

djsteph commented:
Hello Be,

Thank you for reviewing the wiki.

I added the options "iCutEnabled" and "fastSeekEnabled" because of the
Shift Lock feature: If Shift Lock is enabled and you move the platter
accidently, ICut or fast seek could be triggered unintentionnaly. The
alternative would be to remove the options "iCutEnabled" and
"fastSeekEnabled" and have one to enable the Shift lock feature.

I agree with the following: "the TapExpandLibrary option could be removed
if pushing the browse encoder is mapped to [Master]. maximize_library." and
"it would be clearer to rename PADLoopButtonPressed to PADLoopButtonHold
and PADSampleButtonPressed to PADSampleButtonHold."

What color are the pads in Manual Loop Mode? When enabled, they are blue.
If no loop start is present in manual loop, the pads leds are "off"

For the microphone section, I copied the text from the Numark user manual,
but did not test it with Mixxx as I don't have a mic to test with (sorry, I
no longer do live performance).

Thanks

Steph

On Wed, Apr 6, 2016 at 12:15 AM, Be <notifications@github.com> wrote:

> Okay, I read through the wiki. I did quite a bit of editing to make the
> language and formatting more concise and easier to navigate.
>
> It's good that you've included some options, but the amount of options is
> kinda overwhelming. Some of them could be removed. IMO the iCutEnabled and
> fastSeekEnabled options are unnecessary. Because these features have to be
> deliberately activated by the user by holding Shift, they do not interfere
> with other functionality. If a user doesn't want to use them, all they have
> to do is not use them. Keeping the options available clutters the script
> and wiki.
> Also the TapExpandLibrary option could be removed if pushing the browse
> encoder is mapped to [Master]. maximize_library. I think that would be a
> better use of it considering that there are load buttons right next to it.
> I think it would be clearer to rename PADLoopButtonPressed to
> PADLoopButtonHold and PADSampleButtonPressed to PADSampleButtonHold.
>
> What color are the pads in Manual Loop Mode?
>
> Is the microphone input mixed with the main output in hardware? Is the
> input available as a sound card input to the computer?
>
> Are there any volume controls available in your OS mixer program?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/905#issuecomment-206111453>
>

Be-ing commented:
Would making iCutEnabled and fastSeekEnabled only disable those behaviors when shift lock is on would make more sense?

Please add that info about the pad LEDs to the wiki.

You don't actually need a microphone to test how the microphone input works. In Mixxx's Sound Hardware preferences, on the input tab, can you select any inputs from the Mixtrack Pro 3? If not, then the mic input is mixed with the main output in hardware and not routed to the computer.

djsteph commented:
I think that disabling shift lock altogether may be a better option. This
is something that was already in the script when I started to modify it.
This would reduce the number of options in the script.
I think it would also prevent unintentionnal actions... e.g. touching a
button and not realizing that shift lock is, resulting in an unexpected
behavior.

I'll adjust the wiki when I have more spare time .... same for mic testing.

Stéphane Morin
Maison: 450 510-0622
Cellulaire: 514 243-5876

On Wed, Apr 6, 2016 at 2:46 PM, Be <notifications@github.com> wrote:

> Would making iCutEnabled and fastSeekEnabled only disable those behaviors
> when shift lock is on would make more sense?
>
> Please add that info about the pad LEDs to the wiki.
>
> You don't actually need a microphone to test how the microphone input
> works. In Mixxx's Sound Hardware preferences, on the input tab, can you
> select any inputs from the Mixtrack Pro 3? If not, then the mic input is
> mixed with the main output in hardware and not routed to the computer.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/mixxxdj/mixxx/pull/905#issuecomment-206510268>
>

djsteph commented:
Hello,

I checked for the mic, and I don't see any mic inputfrom the controller. So
mic is mixed with the main output in hardware and not routed to the computer

Stéph

On Wed, Apr 6, 2016 at 3:01 PM, Stephane Morin <steph@smorin.com> wrote:

> I think that disabling shift lock altogether may be a better option. This
> is something that was already in the script when I started to modify it.
> This would reduce the number of options in the script.
> I think it would also prevent unintentionnal actions... e.g. touching a
> button and not realizing that shift lock is, resulting in an unexpected
> behavior.
>
> I'll adjust the wiki when I have more spare time .... same for mic testing.
>
> Stéphane Morin
> Maison: 450 510-0622
> Cellulaire: 514 243-5876
>
> On Wed, Apr 6, 2016 at 2:46 PM, Be <notifications@github.com> wrote:
>
>> Would making iCutEnabled and fastSeekEnabled only disable those behaviors
>> when shift lock is on would make more sense?
>>
>> Please add that info about the pad LEDs to the wiki.
>>
>> You don't actually need a microphone to test how the microphone input
>> works. In Mixxx's Sound Hardware preferences, on the input tab, can you
>> select any inputs from the Mixtrack Pro 3? If not, then the mic input is
>> mixed with the main output in hardware and not routed to the computer.
>>
>> —
>> You are receiving this because you authored the thread.
>> Reply to this email directly or view it on GitHub
>> <https://github.com/mixxxdj/mixxx/pull/905#issuecomment-206510268>
>>
>
>

Be-ing commented:
It seems that double tapping the Shift button isn't something that would be easy to do accidentally, so I don't think making an option to disable would be very useful. It also seems to me that use case when shift lock would be most useful would be for using iCut or fast track searching.

djsteph commented:
You are correct Be! That is when shift lock is most useful.

So how do we conclude? To me, the options are available at the time of
set-up then the user forgets about it. There may be a lot, but having used
the controller, I felt that they could useful. The mapping has been
downloaded many times and got no feedback regarding the options.

Envoyé de mon iPad

Le 6 avr. 2016 à 23:20, Be <notifications@github.com> a écrit :

It seems that double tapping the Shift button isn't something that would be
easy to do accidentally, so I don't think making an option to disable would
be very useful. It also seems to me that use case when shift lock would be
most useful would be for using iCut or fast track searching.

—
You are receiving this because you authored the thread.
Reply to this email directly or view it on GitHub
<https://github.com/mixxxdj/mixxx/pull/905#issuecomment-206679174>

Be-ing commented:
Okay, I guess leave those options as they are. I edited the "Configuration options" and "Platter/Jog Wheel" sections of the wiki to clarify their purpose.

Leave a comment when you've made the other changes

djsteph committed:
Update based on Be 's feedback

*            - Renamed user options: PADLoopButtonPressed to
PADLoopButtonHold
 *            - Renamed user options:
PADSampleButtonPressed to PADSampleButtonHold
 *            -
TapExpandLibrary moved from Tap button to Browse Button push
 *
- Linked printComments to debug value of init function: The debugging
parameter is set to 'true' if the user specified the --mididebug
parameter on the command line
 *            - Cleaned
NumarkMixtrack3.shutdown function

djsteph commented:
I have pushed a new commit with requested changes. I'll update the wiki when I get a chance to repost the mapping on the forum. If I do it now, it will be confusing.

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -1343,13 +1328,55 @@ NumarkMixtrack3.PlayButton = function(channel, control, value, status, group) {
+    LibraryGroup ="[Master]";
+    LibraryCommand = "maximize_library";
+    }
+    
```
Considering this code only works differently for the Dark Metal skin, I think it would make more sense to change the `skin` option to a boolean called `DarkMetal`.

Be-ing replied:
Other than this little thing, looks good :)

Be-ing commented:
I notice there are 4 empty lines at the end of the JS file. May as well delete the extra 3.

Be-ing commented:
Btw, it would be easier to make sense of the changelog if someone looked back on this in the future if you split the changes into smaller commits. GUI git clients can be helpful for this. Personally I prefer [Git Cola](http://git-cola.github.io/).

djsteph commented:
Hello Be, 

You'll have to give me an exemple, I don't understand what you mean by 

> it would be easier to make sense of the changelog if someone looked back on this in the future if you split the changes into smaller commits

I tried to provide enought details without going to line number changes.

Be-ing commented:
Yes, you have provided enough details. But just looking at the names of the commits, "Update based on Be 's feedback" doesn't provide much info at a glance; a reader would have to read the extended commit message to see what changed. PRs #828, #902, and #912 are good examples with smaller commits. It's not a big deal, so if you find it too much trouble, don't worry about it.

djsteph committed:
Changed skin option to use boolean if Dark Metal skin is used

- changed skin option to use boolean for DarkMetalSkin. It requires
different code to expand library view.
- removed trailing empty lines at end of script

djsteph commented:
Here you go, this should be good to go now :). If all is Ok, the wiki will need updating of options changes, changes to Browse and tap button and include color of LEDs of Pad buttons when in manual mode

Be-ing commented:
JSHint gives one small nitpicky warning:
Numark-Mixtrack-3-scripts.js: line 1749, col 35, A leading decimal point can be confused with a dot: '.75'.

So change `.75` to `0.75`.

Be-ing commented on res/controllers/Numark-Mixtrack-3-scripts.js:
```diff
@@ -1334,32 +1336,17 @@ NumarkMixtrack3.BrowseButton = function(channel, control, value, status, group)
-    LibraryGroup ="[Master]";
-    LibraryCommand = "maximize_library";
-    }
+    if (DarkMetalSkin){
```
The alignment is off on GitHub's web rendering of this file. It seems that's because you've mixed tabs and spaces here. The rest of the file correctly uses 4 spaces to indent. Use that here too.

Also, IMO it's clearer to read to put a space between `)` and `{` for conditional statements. It looks like there are a few places that you didn't put a space between `if`/`for` and an opening `(` too. Some simple find and replace operations should take care of those.

Be-ing commented on res/controllers/Numark-Mixtrack-3.midi.xml:
```diff
@@ -1,5 +1,5 @@
+<!-- Numark Mixtrack 3 and Pro 3 MIDI preset v1.3 -->
 <MixxxControllerPreset mixxxVersion="1.12.0+" schemaVersion="1">
     <info>
 	<name>Numark Mixtrack Pro 3</name>
```
Change this to `Numark Mixtrack (Pro) 3` so users with the non-Pro version know to use this mapping.

djsteph committed:
Minor changes - Jhint review and readability

- line 1749, change .75 to 0.75 in var gammaOutputRange
- added spacing in numerous place for easier reading,
- changed tabs to spaces

djsteph commented:
Thank you Be for taking the time to review this mapping.What is the next step for this mapping to be merge in the master branch for the next release?

rryan commented on res/controllers/Numark-Mixtrack-3.midi.xml:
```diff
@@ -0,0 +1,918 @@
+	<name>Numark Mixtrack (Pro) 3</name>
+	<author>Stéphane Morin</author>
+	<description>The Numark Mixtrack 3 and Numark Mixtrack Pro 3 are the same controller except that the Pro version has an integrated sound card.</description>
+        <wiki>http://www.mixxx.org/wiki/doku.php/numark_mixtrack_pro_3</wiki>
```
Is there a forum post for this mapping yet? Would you like to make a placeholder forum post and put it here? 
The syntax is:

```<forums>url</forums>``` 
make sure to escape the ```&``` as ```&amp;```

rryan commented:
> Thank you Be for taking the time to review this mapping.What is the next step for this mapping to be merge in the master branch for the next release?

Thanks for doing the review @Be-ing  -- I'll merge this but had a quick question about adding a forum link.

djsteph committed:
Added link to Mixtrack Pro 3 forum in XML file

Added link to Mixtrack Pro 3 forum in XML file

djsteph commented:
The link is added in the XML file
Thanks

rryan commented:
:+1:
